### ISSUE TITLE
[IPRoute] Filter args not honored since pyroute2>0.8.1 

### ISSUE THREAD

**Issue by karelyatin** [svinota/pyroute2#1225]
In OpenStack neutron we have a periodic job running with pyroute2 master(currently at commit e106ce4b9084d471d5b92326783dfd08d1d84d47) and since the merge of thread-unsafe branch in master the job started to fail https://bugs.launchpad.net/neutron/+bug/2092567

The failure is related to ip route show not honoring filter flags.

Reproducer:-
`from pyroute2 import IPRoute`
`ip = IPRoute()`
`routes = ip.route('show', table=255)`
`print("Total routes: %d, Routes List: %s" % (len(routes), [route['table'] for route in routes]))`

With master branch it returns:-
Total routes: 20, Routes List: [254, 254, 254, 254, 254, 254, 254, 255, 255, 255, 255, 255, 254, 254, 254, 254, 255, 255, 255, 255]

With 0.8.1 it returns:-
Total routes: 5, Routes List: [255, 255, 255, 255, 255]

Similar applies to other args like family.
`from pyroute2 import IPRoute`
`ip = IPRoute()`
`routes = ip.route('show', table=255, family=1)`
`print("Total routes: %d, Routes List: %s" % (len(routes), [route['table'] for route in routes]))`

With master it returns:-
Total routes: 20, Routes List: [254, 254, 254, 254, 254, 254, 254, 255, 255, 255, 255, 255, 254, 254, 254, 254, 255, 255, 255, 255]

With 0.8.1 it returns:-
Total routes: 0, Routes List: []

with strict_check it slightly differs.
`from pyroute2 import IPRoute`
`ip = IPRoute(strict_check=True)`
`routes = ip.route('show',table=255, family=1)`
`print("Total routes: %d, Routes List: %s" % (len(routes), [route['table'] for route in routes]))`

With master branch it returns:-
Total routes: 9, Routes List: [255, 255, 255, 255, 255, 255, 255, 255, 255]

With 0.8.1 it returns:-
Total routes: 0, Routes List: []

**Comment by svinota** [comment]
Thanks for reporting! Fixing

**Comment by svinota** [comment]
Should be fixed now.

**Comment by karelyatin** [comment]
<< Should be fixed now.

Thanks @svinota for the fix, but now tests starting failing differently after the latest patch. 11 tests still failing, earlier 15 were failing without the fix[1].
3 failing with:- AttributeError: '_SocketWrapper' object has no attribute 'type'
7 failing with:- ValueError: must specify kind for non-dump commands
1 failing with:-testtools.matchers._impl.MismatchError: 1024 != None

[1] https://f6979cd09b635ec823fd-6d94365b9652dc6406843da38342bcca.ssl.cf1.rackcdn.com/938641/2/check/neutron-functional-with-pyroute2-master/42e2297/testr_results.html

**Comment by svinota** [comment]
Thanks! And thank you a lot for your CI as an additional source for compatibility checks.

I'm going to go through the CI errors and will update you here with the resolution. Separate issues will be fixed in separate PRs, but all will be linked here just for convenience.

Should I do something to trigger the CI from my side, or it's a periodic job so it's enough to merge to the master branch here?

**Comment by karelyatin** [comment]
Should I do something to trigger the CI from my side, or it's a periodic job so it's enough to merge to the master branch here?
Yes it's periodic job runs daily around 02:00 UTC. But can also be triggered by posting "recheck" in https://review.opendev.org/c/openstack/neutron/+/938641 after fixes land in master branch.

**Comment by karelyatin** [comment]
Thanks @svinota with latest fixes, it's now down to 4 failures https://876da9fab577c3dfa969-cde9e85112ed24cd4a9fa6f7c11da08d.ssl.cf2.rackcdn.com/periodic/opendev.org/openstack/neutron/master/neutron-functional-with-pyroute2-master/5ab36e8/testr_results.html

3 failing with:- AttributeError: '_SocketWrapper' object has no attribute 'type'
1 failing with:-testtools.matchers._impl.MismatchError: 1024 != None

**Comment by svinota** [comment]
Yep. The `_SocketWrapper` fix is on the way, while the latter issue needs some investigation.

@karelyatin the socket wrapper code was written initially to make thread-oriented code compatible with eventlet. Since then, a lot of things has happened, and recently the library core was completely rewritten using asyncio, which may affect possible code design. I'm going to provide a compatibility layer decent enough to support existing legacy integrations, but maybe — I'm not sure — there are ways to simplify the things and drop the prehistorical wrapper code. But for that I will need assistance from your team. Probably it worth joining some of the IRC meetings? What would you say?

**Comment by karelyatin** [comment]
@svinota yes it would be great. Our weekly meeting happens as per schedule https://meetings.opendev.org/#Neutron_Team_Meeting i.e Every Tuesday 14:00 UTC over channel #openstack-neutron on OFTC network. Whenever you available you can add the topic to meeting on demand agenda at https://wiki.openstack.org/wiki/Network/Meetings#On_Demand_Agenda

So if you could describe what's all needed and how much efforts it requires to move away from legacy pyroute2 bits, may be over a  launchpad bug at https://bugs.launchpad.net/neutron/ or you can brief first over the IRC meeting to start with?

**Comment by svinota** [comment]
@karelyatin I gonna prepare some docs this weekend, so until Tuesday I will be able to shortly describe the current changes and the legacy code status.

To be honest, I have no idea how the new pyroute2 core may affect the neutron code, that's why I need some help from your side — just to review the current changes and create a possible roadmap, or to leave everything as is. As for my side, I try to keep legacy wrappers working as long as I can (not sure it's the best strategy, though), so to keep it compatible with the old architecture and not to break projects that are using the library (e.g. I'm preparing now a wrapper of a totally deprecated & dropped pyroute2 module for the kuryr project, so the next release will not break it, at least immediately :))

**Comment by karelyatin** [comment]
Thanks @svinota with latest get qdisc fix job is green now https://zuul.openstack.org/build/9f8daccbbb6c45d6846acd9772fbe483

**Comment by svinota** [comment]
@karelyatin good, so I close this issue.

Let me know if there will be further errors.

For some obscure reason I can not login to the wiki — after ubuntu one login I get an "invalid token error" on the wiki — so I can not update the meeting agenda. But I plan to attend the meeting anyway, just in case.
